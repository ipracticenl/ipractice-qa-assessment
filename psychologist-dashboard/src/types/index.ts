// Base types
export interface User {
  id: number;
  name: string;
  type: "psychologist" | "client";
}

// Time range base type
export interface TimeRange {
  id: string;
  from: Date;
  to: Date;
}

// Client DTOs
export interface CreateClientDto {
  name: string;
  initialPsychologistIds: number[];
}

export interface ClientSummaryDto {
  id: number;
  name: string;
}

export interface ClientDetailsDto {
  id: number;
  name: string;
  assignedPsychologists: number[];
  bookedAppointments: Appointment[];
}

// Psychologist DTOs
export interface CreatePsychologistDto {
  name: string;
  initialClients: number[];
}

export interface PsychologistSummaryDto {
  id: number;
  name: string;
}

export interface PsychologistDetailsDto {
  id: number;
  name: string;
  assignedClients: number[];
  availableTimeSlots: AvailableTimeSlot[];
  bookedAppointments: BookedAppointment[];
}

// Booking DTOs
export interface CreateBookingDto {
  psychologistId: number;
  availableTimeSlotId: string;
  comment?: string;
}

export interface ClientDto {
  clientId: number;
}

export interface TimeSlotDto {
  from: Date;
  to: Date;
}

// Model types
export interface AvailableTimeSlot extends TimeRange {
  // Inherits id, from, to from TimeRange
}

export interface Appointment extends TimeRange {
  psychologistId: number;
  comment?: string;
}

export interface BookedAppointment extends TimeRange {
  clientId: number;
}

// Legacy types for backward compatibility
export interface Psychologist extends User {
  type: "psychologist";
  clients: string[];
}

export interface Client extends User {
  type: "client";
  psychologist?: string;
}

export interface Availability {
  id: string;
  psychologistId: number;
  startTime: Date;
  endTime: Date;
  isBooked: boolean;
  clientId?: string;
  clientName?: string;
}
