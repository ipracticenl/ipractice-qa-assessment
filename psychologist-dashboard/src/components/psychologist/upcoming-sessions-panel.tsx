import type { FC } from "react";
import { Card, CardHeader, CardTitle, CardContent } from "../ui/card";
import {
  TableHeader,
  TableRow,
  TableHead,
  TableBody,
  TableCell,
  Table,
} from "../ui/table";
import { BookedAppointment } from "@/types";
import { UpcomingSessionsRow } from "./upcoming-sessions-row";

interface UpcomingSessionsPanelProps {
  bookedSessions: BookedAppointment[];
  onCancelSession: (sessionId: string, clientId: number) => void;
}

export const UpcomingSessionsPanel: FC<UpcomingSessionsPanelProps> = ({
  bookedSessions,
  onCancelSession,
}) => {
  return (
    <Card data-testid="upcoming-sessions-panel">
      <CardHeader>
        <CardTitle className="flex items-center gap-2">
          <span className="w-5 h-5 bg-blue-100 rounded-full flex items-center justify-center text-blue-600 text-xs font-bold">
            {bookedSessions.length}
          </span>
          Upcoming Sessions
        </CardTitle>
      </CardHeader>
      <CardContent>
        <div className="overflow-x-auto">
          <Table>
            <TableHeader>
              <TableRow>
                <TableHead>Date & Time</TableHead>
                <TableHead>Client</TableHead>
                <TableHead>Actions</TableHead>
              </TableRow>
            </TableHeader>
            <TableBody>
              {bookedSessions.length === 0 ? (
                <TableRow>
                  <TableCell
                    colSpan={3}
                    className="text-center py-8 text-gray-500"
                  >
                    No upcoming sessions scheduled
                  </TableCell>
                </TableRow>
              ) : (
                bookedSessions.map((session) => (
                  <UpcomingSessionsRow
                    key={session.id}
                    session={session}
                    onCancelSession={onCancelSession}
                  />
                ))
              )}
            </TableBody>
          </Table>
        </div>
      </CardContent>
    </Card>
  );
};
