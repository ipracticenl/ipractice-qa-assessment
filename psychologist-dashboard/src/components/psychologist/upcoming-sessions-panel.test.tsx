import { render, screen } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { UpcomingSessionsPanel } from "./upcoming-sessions-panel";
import { BookedAppointment } from "@/types";

// Mock the UpcomingSessionsRow component
vi.mock("./upcoming-sessions-row", () => ({
  UpcomingSessionsRow: ({ session, onCancelSession }: any) => (
    <div data-testid={`session-row-${session.id}`}>
      <span>Session {session.id}</span>
      <button onClick={() => onCancelSession(session.id, session.clientId)}>
        Cancel
      </button>
    </div>
  ),
}));

// Mock data
const mockSessions: BookedAppointment[] = [
  {
    id: "session-1",
    from: new Date("2024-01-15T10:00:00"),
    to: new Date("2024-01-15T11:00:00"),
    clientId: 1,
  },
  {
    id: "session-2",
    from: new Date("2024-01-16T14:00:00"),
    to: new Date("2024-01-16T15:00:00"),
    clientId: 2,
  },
];

describe("UpcomingSessionsPanel", () => {
  const mockOnCancelSession = vi.fn();

  describe("Snapshot Tests", () => {
    it("should match snapshot when no sessions exist", () => {
      // Given: No sessions are provided
      const emptySessions: BookedAppointment[] = [];

      // When: The component is rendered with no sessions
      const { container } = render(
        <UpcomingSessionsPanel
          bookedSessions={emptySessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should match the snapshot
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot when sessions exist", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered with sessions
      const { container } = render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should match the snapshot
      expect(container).toMatchSnapshot();
    });
  });

  describe("Empty State", () => {
    it("should display empty state message when no sessions exist", () => {
      // Given: No sessions are provided
      const emptySessions: BookedAppointment[] = [];

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={emptySessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The empty state message should be displayed
      expect(
        screen.getByText("No upcoming sessions scheduled")
      ).toBeInTheDocument();
    });

    it("should show count as 0 when no sessions exist", () => {
      // Given: No sessions are provided
      const emptySessions: BookedAppointment[] = [];

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={emptySessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The count should be displayed as 0
      expect(screen.getByText("0")).toBeInTheDocument();
    });
  });

  describe("Sessions Display", () => {
    it("should display sessions when they exist", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: Session rows should be displayed
      expect(screen.getByTestId("session-row-session-1")).toBeInTheDocument();
      expect(screen.getByTestId("session-row-session-2")).toBeInTheDocument();
    });

    it("should show correct count in title", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The count should be displayed correctly
      expect(screen.getByText("2")).toBeInTheDocument();
    });

    it("should pass correct props to UpcomingSessionsRow components", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The mocked row components should be rendered with session data
      expect(screen.getByText("Session session-1")).toBeInTheDocument();
      expect(screen.getByText("Session session-2")).toBeInTheDocument();
    });
  });

  describe("Component Structure", () => {
    it("should render the main panel with correct test ID", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The main panel should be rendered with correct test ID
      expect(screen.getByTestId("upcoming-sessions-panel")).toBeInTheDocument();
    });

    it("should render table headers correctly", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The table headers should be displayed
      expect(screen.getByText("Date & Time")).toBeInTheDocument();
      expect(screen.getByText("Client")).toBeInTheDocument();
      expect(screen.getByText("Actions")).toBeInTheDocument();
    });

    it("should display the correct title", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The title should be displayed
      expect(screen.getByText("Upcoming Sessions")).toBeInTheDocument();
    });

    it("should have correct CSS classes for styling", () => {
      // Given: Sessions are provided
      const sessions = mockSessions;

      // When: The component is rendered
      const { container } = render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The count badge should have correct styling classes
      const countBadge = container.querySelector(".bg-blue-100.text-blue-600");
      expect(countBadge).toBeInTheDocument();
    });
  });

  describe("Edge Cases", () => {
    it("should handle single session correctly", () => {
      // Given: Only one session is provided
      const singleSession = [mockSessions[0]];

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={singleSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: Only one session row should be displayed
      expect(screen.getByTestId("session-row-session-1")).toBeInTheDocument();
      expect(
        screen.queryByTestId("session-row-session-2")
      ).not.toBeInTheDocument();
      expect(screen.getByText("1")).toBeInTheDocument();
    });

    it("should handle large number of sessions", () => {
      // Given: Many sessions are provided
      const manySessions = Array.from({ length: 10 }, (_, index) => ({
        id: `session-${index + 1}`,
        from: new Date(`2024-01-${15 + index}T10:00:00`),
        to: new Date(`2024-01-${15 + index}T11:00:00`),
        clientId: index + 1,
      }));

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={manySessions}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: All sessions should be displayed
      expect(screen.getByText("10")).toBeInTheDocument();
      expect(screen.getByText("Session session-1")).toBeInTheDocument();
      expect(screen.getByText("Session session-10")).toBeInTheDocument();
    });
  });

  describe("Props Validation", () => {
    it("should pass onCancelSession function to child components", () => {
      // Given: Sessions are provided and a mock cancel function
      const sessions = [mockSessions[0]];
      const mockCancelFunction = vi.fn();

      // When: The component is rendered
      render(
        <UpcomingSessionsPanel
          bookedSessions={sessions}
          onCancelSession={mockCancelFunction}
        />
      );

      // Then: The cancel button should be present and functional
      const cancelButton = screen.getByText("Cancel");
      expect(cancelButton).toBeInTheDocument();
    });
  });
});
