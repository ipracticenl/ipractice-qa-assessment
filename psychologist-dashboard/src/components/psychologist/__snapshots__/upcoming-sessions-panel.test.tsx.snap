// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpcomingSessionsPanel > Snapshot Tests > should match snapshot when no sessions exist 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm"
    data-testid="upcoming-sessions-panel"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <div
        class="text-2xl font-semibold leading-none tracking-tight flex items-center gap-2"
      >
        <span
          class="w-5 h-5 bg-blue-100 rounded-full flex items-center justify-center text-blue-600 text-xs font-bold"
        >
          0
        </span>
        Upcoming Sessions
      </div>
    </div>
    <div
      class="p-6 pt-0"
    >
      <div
        class="overflow-x-auto"
      >
        <div
          class="relative w-full overflow-auto"
        >
          <table
            class="w-full caption-bottom text-sm"
          >
            <thead
              class="[&_tr]:border-b"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Date & Time
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Client
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center py-8 text-gray-500"
                  colspan="3"
                >
                  No upcoming sessions scheduled
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`UpcomingSessionsPanel > Snapshot Tests > should match snapshot when sessions exist 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm"
    data-testid="upcoming-sessions-panel"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <div
        class="text-2xl font-semibold leading-none tracking-tight flex items-center gap-2"
      >
        <span
          class="w-5 h-5 bg-blue-100 rounded-full flex items-center justify-center text-blue-600 text-xs font-bold"
        >
          2
        </span>
        Upcoming Sessions
      </div>
    </div>
    <div
      class="p-6 pt-0"
    >
      <div
        class="overflow-x-auto"
      >
        <div
          class="relative w-full overflow-auto"
        >
          <table
            class="w-full caption-bottom text-sm"
          >
            <thead
              class="[&_tr]:border-b"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              >
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Date & Time
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Client
                </th>
                <th
                  class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
            >
              <div
                data-testid="session-row-session-1"
              >
                <span>
                  Session 
                  session-1
                </span>
                <button>
                  Cancel
                </button>
              </div>
              <div
                data-testid="session-row-session-2"
              >
                <span>
                  Session 
                  session-2
                </span>
                <button>
                  Cancel
                </button>
              </div>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
